packages/node-core/src/indexer/storeModelProvider/model/model.ts

```diff
@@ -0,0 +1,199 @@
+// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
+// SPDX-License-Identifier: GPL-3.0
+
+import {FieldsExpression, GetOptions} from '@subql/types-core';
+import {Op, Model, ModelStatic, Transaction, CreationAttributes, Sequelize} from '@subql/x-sequelize';
+import {Fn} from '@subql/x-sequelize/types/utils';
+import _, {cloneDeep} from 'lodash';
+import {getFullOptions, operatorsMap} from './utils';
+
+export type BaseEntity = {id: string; __block_range?: (number | null)[] | Fn};
+
+export interface IModel<T extends BaseEntity> {
+  get(id: string, tx?: Transaction): Promise<T | undefined>;
+
+  getByFields(filters: FieldsExpression<T>[], options: GetOptions<T>, tx?: Transaction): Promise<T[]>;
+
+  set: (id: string, data: T, blockHeight: number, tx?: Transaction) => Promise<void>;
+  bulkCreate(data: T[], blockHeight: number, tx?: Transaction): Promise<void>;
+  bulkUpdate(data: T[], blockHeight: number, fields?: string[], tx?: Transaction): Promise<void>;
+
+  bulkRemove(ids: string[], blockHeight: number, tx?: Transaction): Promise<void>;
+}
+
+// All operations must be carried out within a transaction.
+export class PlainModel<T extends BaseEntity = BaseEntity> implements IModel<T> {
+  // Record the data that was first created in the current transaction.
+  private currentTxCreatedRecord: Record<string, T> = {};
```

Comment:
I think this should be avoided and be done on a DB level, it should be much simpler and less prone to bugs. Its probably just a matter of updating the hooks.

This also lacks an explanation for its purpose.